src/libraries/System.Collections/src/System/Collections/Generic/Queue.cs

```diff
@@ -386,6 +386,64 @@ public void TrimExcess()
             }
         }
 
+        /// <summary>
+        /// Ensures that the capacity of this Queue is at least the specified <paramref name="capacity"/>.
+        /// </summary>
+        /// <param name="capacity">The minimum capacity to ensure</param>
+        public int EnsureCapacity(int capacity)
+        {
+            if (capacity < 0)
+            {
+                throw new ArgumentOutOfRangeException(nameof(capacity), capacity, SR.ArgumentOutOfRange_NeedNonNegNum);
+            }
+
+            if (_array.Length < capacity)
+            {
+                int newCapacity = (int)((long)(_array.Length == 0 ? 1 : _array.Length) * (long)GrowFactor / 100);
+                while (newCapacity < capacity)
+                {
+                    newCapacity = (int)((long)newCapacity * (long)GrowFactor / 100);
+                }
+
+                // MaxArrayLength is defined in Array.MaxArrayLength and in gchelpers in CoreCLR.
+                // It represents the maximum number of elements that can be in an array where
+                // the size of the element is greater than one byte; a separate, slightly larger constant,
+                // is used when the size of the element is one.
+                const int MaxArrayLength = 0x7FEFFFFF;
+                if ((uint)newCapacity > MaxArrayLength)
+                {
+                    newCapacity = MaxArrayLength;
+                    if (newCapacity < capacity) newCapacity = capacity;
+                }
+
+                if (newCapacity < _array.Length + MinimumGrow)
+                {
+                    newCapacity = _array.Length + MinimumGrow;
+                }
+
+                T[] newArray = new T[newCapacity];
```

Comment:
This is duplicating the logic of the SetCapacity method. Why not just call the method here?

Reply:
Because [comment](https://github.com/dotnet/runtime/pull/47149#issuecomment-767283712), there is a little difference.

Reply:
I think we should avoid duplicating the logic. While it would be nice to eliminate redundant checks as well, it's probably not as important since the dominant factor is the array resizing.

Reply:
Fixed.